src/pip/_internal/operations/prepare.py

```diff
@@ -480,19 +480,7 @@ def _complete_partial_requirements(
             logger.debug("Downloading link %s to %s", link, filepath)
             req = links_to_fully_download[link]
             req.local_file_path = filepath
-            # TODO: This needs fixing for sdists
-            # This is an emergency fix for #11847, which reports that
-            # distributions get downloaded twice when metadata is loaded
-            # from a PEP 658 standalone metadata file. Setting _downloaded
-            # fixes this for wheels, but breaks the sdist case (tests
-            # test_download_metadata). As PyPI is currently only serving
-            # metadata for wheels, this is not an immediate issue.
-            # Fixing the problem properly looks like it will require a
-            # complete refactoring of the `prepare_linked_requirements_more`
-            # logic, and I haven't a clue where to start on that, so for now
-            # I have fixed the issue *just* for wheels.
-            if req.is_wheel:
-                self._downloaded[req.link.url] = filepath
+            self._downloaded[req.link.url] = filepath
```

Comment:
cc @pfmoore: this commit, which just fixes the test cases, fixes this code path. Thanks so much for providing all this context here and in the linked issue, it made things much easier to figure out.

Reply:
I'm not sure I follow. It looked to me like the problem was that the code can't handle an index providing metadata for a sdist - it's very hard for me to test this as PyPI doesn't do that, but certainly it's a case pip *should* support.

I'm not sure what `pip download` vs `pip install` has to do with the issue - both should use metadata if it's available to speed up the resolve. And I'll be honest, the tests were pretty incomprehensible to me, so if they *were* incorrect, they need better comments to explain to future users what they are trying to achieve and how!

Reply:
> It looked to me like the problem was that the code can't handle an index providing metadata for a sdist - it's very hard for me to test this as PyPI doesn't do that, but certainly it's a case pip should support.

On this point, you're absolutely right and I've just spent quite a bit of time to figure it out. I thought I had fixed it in this diff, but it turns out that by changing the tests in this commit to use `install --dry-run` instead of `download`, I was actually avoiding ever hitting this line. I have figured out the right way to handle this in the meantime, which I will split out into a separate diff.

> And I'll be honest, the tests were pretty incomprehensible to me,

On this point, I am confused. The tests in this commit (https://github.com/pypa/pip/pull/12186/commits/7ad477ed2cd99d4980a58dc0de4f599e8e637273) have an absolutely gratuitous amount of comments and docstrings to explain what is being simulated, including making entire separate classes to generate fake data: https://github.com/pypa/pip/blob/7ad477ed2cd99d4980a58dc0de4f599e8e637273/tests/functional/test_install_metadata.py#L40-L54

I understand that the *code* to implement metadata-only dists is extremely hard to follow, and that's one of the goals of this change, to simplify it. But your use of the term "incomprehensible" is confusing to me regarding these tests which seem to have quite a few comments. I'm not taking offense, I just want to be clear: am I misunderstanding you?

Reply:
Created #12197 which adds regression tests against double-downloading anything!